torch/csrc/jit/passes/onnx/function_extraction.cpp

```diff
@@ -0,0 +1,1098 @@
+
+  const std::string module_class_name(
+      func_ctx.scope_key_->name().toUnqualString());
+  auto pos = module_class_name.rfind(".");
+  TORCH_INTERNAL_ASSERT(pos != std::string::npos);
+
+  auto construct_unique_module_name = [&](std::string module_name) {
+    auto module_name_variant = module_variant_count_.find(module_name);
+    if (module_name_variant != module_variant_count_.end()) {
+      module_name += ("." + std::to_string(module_name_variant->second));
+    } else {
+      module_variant_count_[module_name] = 0;
+    }
+    module_variant_count_[module_name]++;
+    return module_name;
+  };
+
+  const auto domain_name = module_class_name.substr(0, pos);
+  const auto func_name =
+      construct_unique_module_name(module_class_name.substr(pos + 1));
+
+  auto func_def_n =
+      CreateFunctionDefNode(func_ctx, graph, domain_name, func_name);
+
+  // create and insert local function node to graph.
+  for (const auto& it : func_ctx.scope_ctxs_) {
+    auto scope = it.first;
+    auto& scope_ctx = *it.second;
+    auto func_n =
+        CreateFunctionNode(func_ctx, scope_ctx, graph, domain_name, func_name);
+
+    std::unordered_set<Node*> old_nodes(
+        scope_ctx.nlist_.begin(), scope_ctx.nlist_.end());
+
+    auto last_n = *scope_ctx.nlist_.rbegin();
+    // replace local function node in parent scopes.
+    for (auto& it : scope_ctxs) {
+      const auto& parent_scope = it.first;
+      auto& parent_ctx = *it.second;
+
+      if (!IsAncestor(parent_scope, scope)) {
+        continue;
+      }
+
+      auto& ctx_nlist = parent_ctx.nlist_;
+      GRAPH_DEBUG(
+          "Replace local function node in parent scope: ",
+          it.first->namesFromRoot(),
+          " nodes to remove: ",
+          old_nodes.size(),
+          " parent total nodes: ",
+          ctx_nlist.size());
+
+      // insert local function node
+      auto last_n_it = std::find(ctx_nlist.begin(), ctx_nlist.end(), last_n);
+      ctx_nlist.insert(last_n_it, func_n);
+
+      // remove replaced nodes from list
+      ctx_nlist.erase(
+          std::remove_if(
+              ctx_nlist.begin(),
+              ctx_nlist.end(),
+              [&old_nodes](Node* n) {
+                return old_nodes.find(n) != old_nodes.end();
+              }),
+          ctx_nlist.end());
+
+      GRAPH_DEBUG("Parent total nodes after remove: ", ctx_nlist.size());
+
+      // refresh inputs/outputs.
+      parent_ctx.PopulateInputsOutputs(param_names_);
+    }
+  }
+
+  for (const auto& it : func_ctx.scope_ctxs_) {
+    auto& scope_ctx = *it.second;
+    // delete replaced nodes in graph.
+    for (auto it = scope_ctx.nlist_.rbegin(); it != scope_ctx.nlist_.rend();) {
+      auto* n = *it;
+      it++;
+      GRAPH_DEBUG("Destroying node ", *n);
+      n->destroy();
+    }
+  }
+}
+
+bool FunctionExtractor::ScopeContext::IsIdenticalFuncion(
+    const ScopeContext& other_ctx) const {
+  // TODO: need to differentiate same function under different inputs.
```

Comment:
If we actually need to handle this at some point, could you add a test demonstrates that this TODO isn't done, then refer to that test case here?